ocaml/jane/doc/extensions/unboxed-types/index.md

```diff
@@ -284,4 +284,144 @@ Here's the list of primitives that currently support `[@layout_poly]`:
 * `%array_safe_set`
 * `%array_unsafe_get`
-* `%array_unsafe_set`
\ No newline at end of file
+* `%array_unsafe_set`
+
+# Using unboxed types in structures
+
+Unboxed types can usually be put in structures, though there are some restrictions.
+
+These structures support unboxed type fields, but have some restrictions on field
+orders:
+  * Records
+  * Constructors
+  
+Unboxed numbers can't be put in these structures:
+  * Constructors with inline record fields
+  * Exceptions
+  * Extensible variant constructors
+  * Top-level fields of modules
+  * Tuples
+  
+There aren't fundamental issues with the structures that lack support. They will
+just take some work to implement.
+
+Here's an example of a record with some unboxed fields. We call such a record 
+a "mixed record".
+
+```ocaml
+type t =
+  { str : string;
+    i : int;
+    f : float#;
+  }
+```
+
+## Restrictions on ordering of fields
+
+The below is written about record fields but equally applies to constructor
+arguments.
+
+Suppose a record contains any unboxed field `fld` whose layout is not `value` or a
+combination of `value` layouts. Then, the following restriction applies: All
+fields occuring after `fld` in the record must be "flat", i.e. the GC can
+skip looking at them. The only options for flat fields are immediates (i.e. things
+represented as ints at runtime) and other unboxed numbers.
+
+The following definition is rejected, as the boxed field `s : string` appears
+after the unboxed float field `f`:
+
+```ocaml
+type t_rejected =
+  { f : float#;
+    s : string;
+  }
+  (* Error: Expected all flat fields after non-value field, f,
+            but found boxed field, s. *)
+```
+
+The only relaxation of the above restriction is records that consists
+solely of `float` and `float#` fields. Any ordering of `float` and `float#`
+fields is permitted. The "flat float record optimization" applies to any
+such record&mdash;all of the fields are stored flat, even the `float` ones
+that will require boxing upon projection. The ordering restriction is relaxed
+in this case to provide a better migration story for all-`float` records
+to which the flat float record optimization currently applies.
+
+```ocaml
+type t_flat_float =
+  { x1 : float;
+    x2 : float#;
+  }
+```
+
+The ordering restriction has to do with the "mixed block" runtime
+representation. Read on for more detail about that.
+
+## Generic operations aren't supported
+
+Some operations built in to the OCaml runtime aren't supported for structures
+containing unboxed types.
+
+These operations aren't supported:
+  * polymorphic comparison and equality
+  * polymorphic hash
+  * marshaling
+  
+These operations raise an exception at runtime, similar to how polymorphic
+comparison raises when called on a function.
+
+You should use ppx-derived versions of these operations instead.
+
+## Runtime representation: mixed blocks
+
+The general principle guiding the restriction on field ordering in the user
```

Comment:
As a general principle, the compiler should not change the user-specified field ordering.